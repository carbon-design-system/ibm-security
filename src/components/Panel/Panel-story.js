/**
 * @file Panel stories.
 * @copyright IBM Security 2019 - 2021
 */

import { Add16 } from '@carbon/icons-react';

import { action } from '@storybook/addon-actions';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import React, { Fragment } from 'react';

import { patterns } from '../../../.storybook';

import { disabled, label } from '../_mocks_';
import { labels } from './_mocks_';

import {
  Button,
  Panel,
  PanelContainer,
  PanelContent,
  PanelController,
} from '../..';

const closeButtonLabel = 'Close';

const panelProps = {
  onClose: action('onClose'),
  onOpen: action('onOpen'),
};

const panelContainerProps = {
  title: label,
  subtitle: label,
  labels,
};

const panelContent =
  'Use the `PanelContent` component to have the appropriate formatting applied to content';

const footerLabel = 'Custom footer';

storiesOf(patterns('Panel'), module).add(
  'Default',
  () => (
    <Fragment>
      <Panel
        {...panelProps}
        render={({ active, handleClose, toggleActive }) => (
          <Fragment>
            <Button onClick={toggleActive}>Primary and secondary action</Button>
            <PanelController active={active}>
              <PanelContainer
                {...panelContainerProps}
                closeButton={{
                  onClick: toggleActive,
                  label: text('closeButton.label', closeButtonLabel),
                }}
                primaryButton={{
                  icon: Add16,
                  iconDescription: text('primaryButton.iconDescription', 'Add'),
                  isDisabled: boolean('primaryButton.isDisabled', !disabled),
                  label: text('primaryButton.label', 'Add'),
                  onClick: handleClose,
                }}
                secondaryButton={{
                  isDisabled: boolean('secondaryButton.isDisabled', !disabled),
                  label: text('secondaryButton.label', 'Cancel'),
                  onClick: handleClose,
                }}
              >
                <PanelContent>{panelContent}</PanelContent>
              </PanelContainer>
            </PanelController>
          </Fragment>
        )}
      />
      <Panel
        {...panelProps}
        render={({ active, toggleActive }) => (
          <Fragment>
            <Button onClick={toggleActive}>Primary action</Button>
            <PanelController active={active}>
              <PanelContainer
                {...panelContainerProps}
                closeButton={{
                  onClick: toggleActive,
                  label: text('closeButton.label', closeButtonLabel),
                }}
                primaryButton={{
                  isDisabled: boolean('primaryButton.isDisabled', !disabled),
                  label: text('primaryButton.label', 'Submit'),
                  onClick: toggleActive,
                }}
              />
            </PanelController>
          </Fragment>
        )}
      />
      <Panel
        {...panelProps}
        render={({ active, toggleActive }) => (
          <Fragment>
            <Button onClick={toggleActive}>Default</Button>
            <PanelController active={active}>
              <PanelContainer
                {...panelContainerProps}
                closeButton={{
                  onClick: toggleActive,
                  label: closeButtonLabel,
                }}
              >
                <PanelContent>{panelContent}</PanelContent>
              </PanelContainer>
            </PanelController>
          </Fragment>
        )}
      />
      <Panel
        {...panelProps}
        render={({ active, toggleActive }) => (
          <Fragment>
            <Button onClick={toggleActive}>{footerLabel}</Button>
            <PanelController active={active}>
              <PanelContainer
                {...panelContainerProps}
                closeButton={{
                  onClick: toggleActive,
                  label: text('closeButton.label', closeButtonLabel),
                }}
                renderFooter={() => <Button size="xl">{footerLabel}</Button>}
              />
            </PanelController>
          </Fragment>
        )}
      />
    </Fragment>
  ),

  {
    info:
      "View the 'Story' addon panel for guidance on using the `renderFooter` prop",
  }
);
