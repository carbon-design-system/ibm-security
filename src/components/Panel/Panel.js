/**
 * @file Panel.
 * @copyright IBM Security 2019
 */

import PropTypes from 'prop-types';
import { Component } from 'react';

/**
 * Panel component.
 * @param {object.<string, *>} props Panel props.
 * @returns {Panel} Panel instance.
 */
export default class Panel extends Component {
  static propTypes = {
    /** @type {Node} Child elements. */
    children: PropTypes.node,

    /** @type {function(...args)} Function to call when the panel closes. */
    onClose: PropTypes.func,

    /** @type {function(...args)} Function to call when the panel opens. */
    onOpen: PropTypes.func,

    /** @type {function(...props)} Render function for the contents of the table. */
    render: PropTypes.func,

    /** @type {bool} specifies whether the Panel should open or not */
    shouldPanelOpen: PropTypes.bool,
  };

  static defaultProps = {
    children: undefined,
    onClose: () => {},
    onOpen: () => {},
    render: undefined,
    shouldPanelOpen: true,
  };

  state = { active: false };

  /**
   * Handles when the panel has been requested to close.
   * @param {Event} event Event object generated from close request.
   * @param {...any} args The arguments passed via render props.
   */
  handleClose = (event, ...args) => {
    event.stopPropagation();
    if (this.state.active) {
      this.props.onClose(event, ...args);
      this.setState({ active: false });
    }
  };

  /**
   * Handles when the panel has been requested to open.
   * @param {Event} event Event object generated from open request.
   * @param {...any} args The arguments passed via render props.
   */
  handleOpen = (event, ...args) => {
    event.stopPropagation();
    if (!this.state.active) {
      this.props.onOpen(event, ...args);
      this.setState({ active: true });
    }
  };

  /**
   * Handles when the panel has been requested to toggle.
   * @param {Event} event Event object generated from toggle request.
   * @param {...any} args The arguments passed via render props.
   */
  toggleActive = (event, ...args) => {
    if (this.state.active) {
      this.props.onClose(event, ...args);
      this.setState({ active: !this.state.active });
    } else if (this.props.shouldPanelOpen) {
      this.props.onOpen(event, ...args);
      this.setState({ active: !this.state.active });
    }
  };

  render() {
    const { children, render } = this.props;
    const renderProps = {
      active: this.state.active,
      handleClose: this.handleClose,
      handleOpen: this.handleOpen,
      toggleActive: this.toggleActive,
    };

    if (render !== undefined) {
      return render(renderProps);
    }
    if (children !== undefined) {
      return children(renderProps);
    }
    return null;
  }
}
