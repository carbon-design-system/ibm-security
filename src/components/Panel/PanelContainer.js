/**
 * @file Panel container.
 * @copyright IBM Security 2019
 */

/* eslint-disable no-useless-computed-key, react/require-default-props */

import Close20 from '@carbon/icons-react/lib/close/20';

import classnames from 'classnames';
import PropTypes, { func } from 'prop-types';
import React, { Component, createRef, Fragment } from 'react';
import { createPortal } from 'react-dom';

import requiredIfGivenPropExists from 'carbon-components-react/lib/prop-types/requiredIfGivenPropExists';
import setupGetInstanceId from 'carbon-components-react/lib/tools/setupGetInstanceId';

import { getComponentNamespace } from '../../globals/namespace';
import * as defaultLabels from '../../globals/nls';

import { isClient, isNode } from '../../globals/utils/capabilities';
import { focusFirstElement, trapTabFocus } from '../../globals/utils/focus';

import Button from '../Button';
import IconButton from '../IconButton';

export const namespace = getComponentNamespace('panel');
const getInstanceId = setupGetInstanceId();

/**
 * Panel container component.
 * @param {Object.<string, *>} props Panel container props.
 * @returns {PanelContainer} Panel container instance.
 */
export default class PanelContainer extends Component {
  constructor(props) {
    super(props);
    if (isClient()) {
      this.container = this.getContainer();
      this.containerClass = `${namespace}__container`;
      this.element = document.createElement('section');

      this.element.classList.add(namespace);

      this.previousFocus = document.activeElement;
      this.rootNode = this.props.rootNode;

      const { className } = this.props;

      if (className) {
        this.element.classList.add(className);
      }
    }
  }

  state = { bodyMargin: 0 };

  componentDidMount() {
    if (isClient()) {
      if (!this.container) {
        this.container = document.createElement('div');
        this.rootNode.appendChild(this.container);
      }

      this.container.appendChild(this.element);
      if (this.container.childElementCount === 1) {
        this.rootNode.classList.toggle(this.containerClass);
      }

      focusFirstElement(this.element);
      this.element.addEventListener('keydown', this.handleKeyPress);

      this.setBodyMargin();
    }
  }

  componentWillUnmount() {
    if (isClient()) {
      this.container.removeChild(this.element);

      if (this.container.childElementCount === 0) {
        this.rootNode.classList.toggle(this.containerClass);
        this.rootNode.removeChild(this.container);
      }

      this.element.removeEventListener('keydown', this.handleKeyPress);
      this.previousFocus.focus();
    }
  }

  /**
   * Retrieves the element to render the panel container in to via a portal.
   * @returns {HTMLElement} The element to render the panel container in to via a portal.
   */
  getContainer = () => {
    if (isClient()) {
      const panels = document.getElementsByClassName(namespace);

      if (panels.length > 0) {
        return panels[0].parentNode;
      }
    }

    return null;
  };

  /**
   * Sets the body margin to match the height of the header for fixed scrolling.
   */
  setBodyMargin() {
    const { current: footer } = this.footer;

    this.setState({
      bodyMargin: {
        top: this.header.current.clientHeight,
        bottom: footer && footer.clientHeight,
      },
    });
  }

  panelInstanceId = `panel-${getInstanceId()}`;
  panelTitleId = `${namespace}__title--${this.panelInstanceId}`;
  panelSubtitleId = `${namespace}__subtitle--${this.panelInstanceId}`;

  footer = createRef();
  header = createRef();

  /**
   * Handles the key press focus trap.
   * @param {SyntheticEvent} event The event fired when a key has been pressed while the panel container is in focus.
   */
  handleKeyPress = event => {
    if (event.key === 'Tab') {
      trapTabFocus(this.element, event);
    } else if (!this.props.disableEscape && event.key === 'Escape') {
      this.props.closeButton.onClick();
    }
  };

  renderPanelContainer = ({
    labels: {
      PANEL_CONTAINER_PRIMARY_BUTTON,
      PANEL_CONTAINER_SECONDARY_BUTTON,
      PANEL_CONTAINER_CLOSE_BUTTON,
    },
  }) => {
    const {
      children,
      closeButton,
      primaryButton,
      renderFooter,
      secondaryButton,
      subtitle,
      title,
      hasScrollingContent,
    } = this.props;

    const hasFooter = renderFooter || primaryButton;

    const ariaLabel = this.props['aria-label'] || title || subtitle;

    const getAriaLabelledBy = title ? this.panelTitleId : this.panelSubtitleId;

    const hasScrollingContentProps = hasScrollingContent
      ? {
          tabIndex: 0,
          role: 'region',
          'aria-label': ariaLabel,
          'aria-labelledby': getAriaLabelledBy,
        }
      : {};

    return (
      <div role="dialog" aria-label={ariaLabel} aria-modal="true">
        <header ref={this.header} className={`${namespace}__header`}>
          {title && (
            <div className={`${namespace}__header__container--title`}>
              <div
                id={this.panelTitleId}
                className={`${namespace}__header--title`}
              >
                {title}
              </div>
              {subtitle && (
                <div
                  id={this.panelSubtitleId}
                  className={`${namespace}__header--subtitle`}
                >
                  {subtitle}
                </div>
              )}
            </div>
          )}
          <IconButton
            id={closeButton.id}
            className={`${namespace}__button--close`}
            label={PANEL_CONTAINER_CLOSE_BUTTON}
            onClick={closeButton.onClick}
            renderIcon={closeButton.icon || Close20}
            tooltip={false}
          />
        </header>
        <section
          className={classnames(`${namespace}__body`, {
            [`${namespace}__body--footer`]: hasFooter,
          })}
          style={{
            marginTop: `${this.state.bodyMargin.top}px`,
            marginBottom: `${this.state.bodyMargin.bottom}px`,
          }}
          {...hasScrollingContentProps}
          aria-labelledby={getAriaLabelledBy}
        >
          {children}
        </section>
        {hasFooter && (
          <footer ref={this.footer} className={`${namespace}__footer`}>
            {renderFooter ? (
              renderFooter()
            ) : (
              <Fragment>
                {secondaryButton && (
                  <Button
                    id={secondaryButton.id}
                    className={`${namespace}__footer__button ${namespace}__footer__button--secondary`}
                    disabled={secondaryButton.isDisabled}
                    iconDescription={secondaryButton.iconDescription}
                    kind="secondary"
                    onClick={secondaryButton.onClick}
                    renderIcon={secondaryButton.icon}
                  >
                    {PANEL_CONTAINER_SECONDARY_BUTTON}
                  </Button>
                )}
                <Button
                  id={primaryButton.id}
                  className={`${namespace}__footer__button`}
                  disabled={primaryButton.isDisabled}
                  iconDescription={primaryButton.iconDescription}
                  onClick={primaryButton.onClick}
                  renderIcon={primaryButton.icon}
                >
                  {PANEL_CONTAINER_PRIMARY_BUTTON}
                </Button>
              </Fragment>
            )}
          </footer>
        )}
      </div>
    );
  };

  render() {
    const { closeButton, primaryButton, secondaryButton, labels } = this.props;

    const componentLabels = {
      ...defaultLabels.labels,
      ...labels,
      ...defaultLabels.filterFalsey({
        PANEL_CONTAINER_PRIMARY_BUTTON:
          (primaryButton && primaryButton.label) || '',
        PANEL_CONTAINER_SECONDARY_BUTTON:
          (secondaryButton && secondaryButton.label) || '',
        PANEL_CONTAINER_CLOSE_BUTTON: (closeButton && closeButton.label) || '',
      }),
    };
    return (
      isClient() &&
      createPortal(
        this.renderPanelContainer({ labels: componentLabels }),
        this.element
      )
    );
  }
}

const buttonType = PropTypes.shape({
  id: PropTypes.string,
  onClick: PropTypes.func,
  label: PropTypes.string,
  isDisabled: PropTypes.bool,
  icon: PropTypes.object,
  iconDescription: PropTypes.string,
});

PanelContainer.propTypes = {
  /** @type {ReactNode} The children of the panel container. */
  children: PropTypes.node,

  /** @type {string} Class name. */
  className: PropTypes.string,

  /** @type {Object<Object>} An object list of close button props. */
  closeButton: buttonType.isRequired,

  /** @type {boolean} Set to true to disable the 'Escape' key from closing the panel. */
  disableEscape: PropTypes.bool,

  /** @type {Object<Object>} An object list of primary button props. */
  primaryButton: buttonType,

  /** @type {Object<Object>} An object list of secondary button props. */
  secondaryButton: buttonType,

  /** @type {string} Child elements. */
  subtitle: PropTypes.node,

  /** @type {string} Child elements. */
  title: PropTypes.node,

  /** @type {function} Footer render prop. */
  renderFooter: func,

  /** @type {string} Root node to attach the panel to. */
  rootNode: isNode() ? PropTypes.instanceOf(Node) : PropTypes.any,

  labels: defaultLabels.propType,

  /**
   * Specify whether the panel contains scrolling content
   */
  hasScrollingContent: PropTypes.bool,

  /**
   * Required props for the accessibility label of the header
   */
  ['aria-label']: requiredIfGivenPropExists(
    'hasScrollingContent',
    PropTypes.string
  ),
};

PanelContainer.defaultProps = {
  children: null,
  className: null,
  disableEscape: false,
  primaryButton: null,
  secondaryButton: undefined,
  subtitle: undefined,
  title: undefined,
  renderFooter: null,
  rootNode: isNode() ? document.body : undefined,
  labels: {},
  hasScrollingContent: false,
};

/* eslint-enable */
