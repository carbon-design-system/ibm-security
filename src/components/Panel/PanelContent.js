/**
 * @file Panel content.
 * @copyright IBM Security 2019
 */

import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { getComponentNamespace } from '../../globals/namespace';

const namespace = getComponentNamespace('panel');

/**
 * Panel content component.
 * @param {Object.<string, *>} props Panel content props.
 * @returns {PanelContent} Panel content instance.
 */
const PanelContent = props => {
  const classes = classnames(`${namespace}__body--content`, props.className);
  return <div className={classes}>{props.children}</div>;
};

PanelContent.defaultProps = {
  children: undefined,
  className: undefined,
};

PanelContent.propTypes = {
  /** @type {ReactNode} Child elements. */
  children: PropTypes.node,

  /** @type {string} Extra classes to add. */
  className: PropTypes.string,
};

export default PanelContent;
