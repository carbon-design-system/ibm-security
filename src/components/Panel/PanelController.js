/**
 * @file Panel controller.
 * @copyright IBM Security 2019
 */

import PropTypes from 'prop-types';

/**
 * Panel controller component.
 * @param {object.<string, *>} props Panel controller props.
 * @returns {PanelController} Panel controller instance.
 */
const PanelController = (props) => (props.active ? props.children : null);

PanelController.defaultProps = {
  active: false,
  children: undefined,
};

PanelController.propTypes = {
  /** @type {boolean} Set to true to render the children. */
  active: PropTypes.bool,

  /** @type {ReactNode} Child elements. */
  children: PropTypes.node,
};

export default PanelController;
