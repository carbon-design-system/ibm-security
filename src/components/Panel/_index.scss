////
/// Panel component.
/// @group panel
/// @copyright IBM Security 2019
////

@import '../Button/index';
@import '../IconButton/index';

@import '../Component/mixins';

@import 'mixins';

@include security--component($name: panel) {
  @include panel;
}
