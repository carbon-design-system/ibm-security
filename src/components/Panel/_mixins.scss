////
/// Panel mixins.
/// @group panel
/// @copyright IBM Security 2019 - 2020
////

@import '@carbon/layout/scss/spacing';
@import '@carbon/layout/scss/mini-unit';
@import '@carbon/type/scss/index';
@import '@carbon/themes/scss/tokens';
@import 'carbon-components/scss/globals/scss/layout';

@import '../../globals/rtl/index';
@import '../../globals/theme/mixins';

@import '../../platform/body/mixins';

@import '../Header/mixins';
@import '../IconButton/mixins';
@import '../Toolbar/mixins';

@mixin panel {
  /// Width.
  /// @type Length
  $panel__sizing__width: carbon--mini-units(
    $count: 40,
  );

  /// Padding.
  /// @type Length
  $panel__spacing__padding: $carbon--spacing-05;

  /// Header width.
  /// @type Length
  $panel__header__sizing__width: $panel__sizing__width -
    ($panel__spacing__padding * 2);

  /// Content border color.
  /// @type Color
  $panel__content__color__border: $ui-04;

  /// Content border sizing.
  /// @type Color
  $panel__content__sizing__border: carbon--mini-units(
    $count: 0.125,
  );

  @include security--theme--cg10__cg100;

  display: flex;
  position: fixed;
  top: 0;
  right: 0;
  height: 100%;
  width: $panel__sizing__width;
  color: $body__color__text;
  background-color: $ui-01;
  flex-direction: column;
  justify-content: space-between;
  z-index: z($layer: modal);

  @include rtl {
    right: unset;
    left: 0;
  }

  &__container #{$header__namespace}__container {
    width: calc(
      #{$header__sizing__width} - #{$toolbar__sizing__width} - #{$panel__sizing__width}
    );

    > #{$header__namespace} {
      width: inherit;
    }
  }

  &__button--close:only-child {
    @include rtl {
      margin-right: auto;
      margin-left: unset;
    }

    margin-left: auto;
  }

  &__header {
    display: flex;
    position: fixed;
    top: 0;
    width: $panel__header__sizing__width;
    padding-top: $panel__spacing__padding;
    padding-bottom: $panel__spacing__padding;
    margin-right: $panel__spacing__padding;
    margin-left: $panel__spacing__padding;
    box-sizing: border-box;
    justify-content: space-between;
    border-bottom: $panel__content__sizing__border solid
      $panel__content__color__border;

    &__container--title {
      width: $panel__header__sizing__width -
        carbon--mini-units($count: $button--icon__sizing__dimensions) -
        $panel__spacing__padding;
    }

    &--title,
    &--subtitle {
      color: $text-01;
      word-break: break-word;
    }

    &--title {
      @include carbon--type-style($name: productive-heading-03);

      display: block;
      margin-bottom: $carbon--spacing-04;
    }

    &--subtitle {
      @include carbon--type-style($name: body-short-01);
    }
  }

  &__body {
    margin-top: carbon--mini-units($count: 11);
    flex-grow: 1;
    overflow-y: auto;

    &--content {
      @include carbon--type-style($name: caption-01);

      padding: $panel__spacing__padding;
      border-bottom: $panel__content__sizing__border solid
        $panel__content__color__border;

      &:last-child {
        border-bottom: none;
      }
    }

    &--footer {
      margin-bottom: $carbon--layout-04;
    }
  }

  &__footer {
    display: flex;
    position: fixed;
    bottom: 0;
    width: $panel__sizing__width;

    &__button {
      max-width: $panel__sizing__width;
      flex-grow: 1;
    }
  }
}
