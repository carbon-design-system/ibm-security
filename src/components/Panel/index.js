/**
 * @file Panel entry point.
 * @copyright IBM Security 2019
 */

export default from './Panel';

export PanelContent from './PanelContent';
