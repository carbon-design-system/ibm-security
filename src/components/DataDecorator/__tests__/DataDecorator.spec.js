/**
 * @file Data decorator tests.
 * @copyright IBM Security 2019
 */

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import renderWithinLandmark from '../../../../config/jest/helpers/renderWithinLandmark';

import { Button, DataDecorator } from '../../..';

import { namespace as panelNamespace } from '../../Panel/Panel';

describe('DataDecorator', () => {
  test('should have no Axe or DAP violations', async () => {
    const { container } = renderWithinLandmark(
      <DataDecorator type="IP" value="10.0.0.0" score={0} href="#" />
    );

    await expect(container).toHaveNoAxeViolations();
    await expect(container).toHaveNoDAPViolations('DataDecorator');
  });

  test('should have no Axe or DAP violations with an open panel', async () => {
    const { container, getByText } = renderWithinLandmark(
      <DataDecorator type="IP" value="10.0.0.0" score={0} href="#" />
    );

    // Click on the data decorator to open the connected panel:
    userEvent.click(getByText(/10.0.0.0/i).closest('button'));

    await expect(container).toHaveNoAxeViolations();
    await expect(container).toHaveNoDAPViolations(
      'DataDecorator with open panel'
    );
  });

  test('should use the value of the data decorator to the panel title', () => {
    const { getByText, queryAllByText } = render(
      <DataDecorator type="IP" value="10.0.0.0" score={0} />
    );
    // Because the panel is closed, we expect only 1 element with
    // the IP address -- the data decorator.
    expect(queryAllByText(/10.0.0.0/i).length === 1);

    // Open the panel:
    userEvent.click(getByText(/10.0.0.0/i).closest('button'));

    // Because the panel is open, we expect to see 2 elements with
    // the IP address -- the data decorator button AND the panel title.
    expect(queryAllByText(/10.0.0.0/i).length === 2);
  });

  test('should invoke open mock when data decorator is clicked to open panel', () => {
    const onOpenMock = jest.fn();
    const { getByText } = render(
      <DataDecorator type="IP" value="10.0.0.0" score={0} onOpen={onOpenMock} />
    );

    userEvent.click(getByText(/10.0.0.0/i).closest('button'));
    expect(onOpenMock).toHaveBeenCalledTimes(1);
  });

  test("should invoke close mock when data decorator's open panel's close button is clicked", () => {
    const onCloseMock = jest.fn();
    const { getByLabelText, getByText } = render(
      <DataDecorator
        type="IP"
        value="10.0.0.0"
        score={0}
        href="#"
        closeButton={{
          label: 'test close',
          onClick: onCloseMock,
        }}
      />
    );
    // Open the panel first:
    userEvent.click(getByText(/10.0.0.0/i).closest('button'));

    userEvent.click(getByLabelText(/test close/i));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  test('should traverse between data decorator and open panel in tab order', async () => {
    const onCloseMock = jest.fn();
    const { getByLabelText, getByText } = render(
      <DataDecorator
        type="IP"
        value="10.0.0.0"
        score={0}
        href="#"
        closeButton={{
          label: 'test close',
          onClick: onCloseMock,
        }}
      >
        <Button>test content button</Button>
      </DataDecorator>
    );

    userEvent.tab();

    // The data decorator:
    expect(getByText(/10.0.0.0/i).closest('button')).toHaveFocus();

    userEvent.tab();

    // Expect the data decorator to still have focus
    // because it is currently the only interactive element:
    expect(getByText(/10.0.0.0/i).closest('button')).toHaveFocus();

    // Simulate a click to open the connected panel.
    userEvent.click(getByText(/10.0.0.0/i).closest('button'));

    // Save off the panel selector.
    // We will need it to set the `focusTrap` option below.
    // See: https://github.com/testing-library/user-event#tabshift-focustrap
    const panel = document.querySelector(`.${panelNamespace}`);

    // Tab into the open panel.
    // Set focus trap.
    userEvent.tab({ focusTrap: panel });

    // The panel's close button:
    expect(getByLabelText(/test close/i)).toHaveFocus();

    // Maintain focus trap in open panel:
    userEvent.tab({ focusTrap: panel });

    // The example button inside the panel:
    expect(getByText(/test content button/i)).toHaveFocus();

    // Maintain focus trap in open panel:
    userEvent.tab({ focusTrap: panel });

    // Panel loop complete.
    // Focus cannot leave panel until it is closed.
    expect(getByLabelText(/test close/i)).toHaveFocus();
  });
});
