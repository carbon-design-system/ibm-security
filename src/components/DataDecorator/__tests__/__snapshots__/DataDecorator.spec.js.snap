// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataDecorator Rendering renders correctly 1`] = `
<DataDecorator
  href="#"
  inline={false}
  labels={Object {}}
  onClose={[Function]}
  onOpen={[MockFunction]}
  primaryButton={
    Object {
      "onClick": [Function],
    }
  }
  renderFooter={null}
  score={0}
  scoreThresholds={
    Array [
      0,
      4,
      7,
      10,
    ]
  }
  type="IP"
  value="10.0.0.0"
>
  <Panel
    onClose={[Function]}
    onOpen={[MockFunction]}
    render={[Function]}
    shouldPanelOpen={true}
  >
    <Decorator
      active={false}
      className=""
      inert={false}
      inline={false}
      noIcon={false}
      onClick={[Function]}
      score={0}
      scoreThresholds={
        Array [
          0,
          4,
          7,
          10,
        ]
      }
      type="IP"
      value="10.0.0.0"
    >
      <button
        className="security--decorator security--decorator--low"
        onClick={[Function]}
      >
        <span
          className="security--decorator__icon"
        >
          <Icon
            className=""
            fillRule="evenodd"
            height={null}
            path="M 8.00,3.00
           C 8.00,3.00 14.00,12.00 14.00,12.00
             14.00,12.00 2.00,12.00 2.00,12.00
             2.00,12.00 8.00,3.00 8.00,3.00 Z"
            renderIcon={null}
            size={16}
            viewBox="0 0 16 16"
            width={null}
          >
            <svg
              aria-hidden="true"
              className="security--icon"
              fillRule="evenodd"
              focusable={false}
              height={16}
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "willChange": "transform",
                }
              }
              viewBox="0 0 16 16"
              width={16}
            >
              <path
                d="M 8.00,3.00
           C 8.00,3.00 14.00,12.00 14.00,12.00
             14.00,12.00 2.00,12.00 2.00,12.00
             2.00,12.00 8.00,3.00 8.00,3.00 Z"
              />
            </svg>
          </Icon>
        </span>
        <span
          className="security--decorator__type"
        >
          IP
        </span>
        <span
          className="security--decorator__value"
        >
          10.0.0.0
        </span>
      </button>
    </Decorator>
    <PanelController
      active={false}
    />
  </Panel>
</DataDecorator>
`;

exports[`DataDecorator Rendering renders the HTML of the node's subtree 1`] = `
<button
  class="security--decorator security--decorator--low"
>
  <span
    class="security--decorator__icon"
  >
    <svg
      aria-hidden="true"
      class="security--icon"
      fill-rule="evenodd"
      focusable="false"
      height="16"
      preserveAspectRatio="xMidYMid meet"
      style="will-change: transform;"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        d="M 8.00,3.00
           C 8.00,3.00 14.00,12.00 14.00,12.00
             14.00,12.00 2.00,12.00 2.00,12.00
             2.00,12.00 8.00,3.00 8.00,3.00 Z"
      />
    </svg>
  </span>
  <span
    class="security--decorator__type"
  >
    IP
  </span>
  <span
    class="security--decorator__value"
  >
    10.0.0.0
  </span>
</button>
`;
